File: src/test/java/com/codeborne/selenide/conditions/VisibleTest.java
package com.codeborne.selenide.conditions;

import com.codeborne.selenide.DriverStub;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.WebElement;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

final class VisibleTest {
  private final Visible condition = new Visible();
  private final WebElement element = mock(WebElement.class);

  @Test
  void negate() {
    assertThat(condition.missingElementSatisfiesCondition()).isFalse();
    assertThat(condition.negate().missingElementSatisfiesCondition()).isTrue();
  }

  @Test
  void name() {
    assertThat(condition.getName()).isEqualTo("visible");
    assertThat(condition.negate().getName()).isEqualTo("not visible");
  }

  @Test
  void satisfied_if_element_is_visible() {
    when(element.isDisplayed()).thenReturn(true);
    assertThat(condition.apply(new DriverStub(), element)).isTrue();
  }

  @Test
  void not_satisfied_if_element_is_invisible() {
    when(element.isDisplayed()).thenReturn(false);
    assertThat(condition.apply(new DriverStub(), element)).isFalse();
  }

  @Test
  void actualValue_invisible() {
    assertThat(condition.actualValue(new DriverStub(), element)).isEqualTo("visible:false");
  }

  @Test
  void actualValue_visible() {
    when(element.isDisplayed()).thenReturn(true);
    assertThat(condition.actualValue(new DriverStub(), element)).isEqualTo("visible:true");
  }
}

File: src/jvmTest/java/com/codeborne/selenide/conditions/VisibleTest.kt
package com.codeborne.selenide.conditions

import com.codeborne.selenide.DriverStub
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.openqa.selenium.WebElement

@ExperimentalCoroutinesApi
internal class VisibleTest {
    private val condition = Visible()
    private val element = Mockito.mock(WebElement::class.java)
    @Test
    fun negate() {
        Assertions.assertThat(condition.missingElementSatisfiesCondition()).isFalse
        Assertions.assertThat(condition.negate().missingElementSatisfiesCondition()).isTrue
    }

    @Test
    fun name() {
        Assertions.assertThat(condition.name).isEqualTo("visible")
        Assertions.assertThat(condition.negate().name).isEqualTo("not visible")
    }

    @Test
    fun satisfied_if_element_is_visible() = runBlockingTest {
        Mockito.`when`(element.isDisplayed).thenReturn(true)
        Assertions.assertThat(condition.apply(DriverStub(), element)).isTrue()
    }

    @Test
    fun not_satisfied_if_element_is_invisible() = runBlockingTest {
        Mockito.`when`(element.isDisplayed).thenReturn(false)
        Assertions.assertThat(condition.apply(DriverStub(), element)).isFalse()
    }

    @Test
    fun actualValue_invisible() = runBlockingTest {
        Assertions.assertThat<Any>(condition.actualValue(DriverStub(), element)).isEqualTo("visible:false")
    }

    @Test
    fun actualValue_visible() = runBlockingTest {
        Mockito.`when`(element.isDisplayed).thenReturn(true)
        Assertions.assertThat<Any>(condition.actualValue(DriverStub(), element)).isEqualTo("visible:true")
    }
}
